File: javascript/test/strings/kmp.test.js
const { search } = require('../../src/strings/kmp');

describe('kmp', () => {
  test('found', () => {
    expect(search('ababcababa', 'aba')).toBe(0);
  });

  test('not found', () => {
    expect(search('hello', 'world')).toBe(-1);
  });

  test('empty pattern', () => {
    expect(search('abc', '')).toBe(0);
  });

  test('null', () => {
    expect(search(null, 'a')).toBe(-1);
  });
});

File: java/src/test/java/com/dsa/examples/strings/KMPTest.java
package com.dsa.examples.strings;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class KMPTest {

    @Test
    void findsPattern() {
        assertEquals(0, KMP.search("ababcababa", "aba"));
    }

    @Test
    void notFound() {
        assertEquals(-1, KMP.search("hello", "world"));
    }

    @Test
    void emptyPattern() {
        assertEquals(0, KMP.search("abc", ""));
    }

    @Test
    void nullText() {
        assertEquals(-1, KMP.search(null, "a"));
    }
}
